src/NHibernate/Type/GenericMapType.cs

```diff
@@ -116,5 +117,12 @@ public override object IndexOf(object collection, object element)
 				.Select(pair => pair.Key)
 				.FirstOrDefault();
 		}
+
+		protected override bool AreCollectionElementsEqual(IEnumerable original, IEnumerable target)
+		{
+			var first = (IDictionary<TKey, TValue>)original;
+			var second = (IDictionary<TKey, TValue>)target;
+			return first.Count == second.Count && first.All(keyValue => second.Contains(keyValue));
```

Comment:
Your are right, that is me who have miss read, sorry. I was believing you were testing only keys with `ContainsKey`.
So this is O(n²) indeed. Well, as the bag `EqualsSnapshot` implementation is, so why not.